File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        
        minHeap = []

        for i in range(len(lists)):
            if lists[i]:
                heapq.heappush(minHeap, (lists[i].val, i, lists[i]))

        dummy = ListNode(-1)
        current = dummy

        while minHeap:
            node = heapq.heappop(minHeap)[2]
            current.next = node
            current = current.next

            if node.next:
                i += 1
                heapq.heappush(minHeap, (node.next.val, i, node.next))

        return dummy.next

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {

        if (lists == null || lists.length == 0) return null;
        
        PriorityQueue<ListNode> minHeap = new PriorityQueue<ListNode>(lists.length, (a,b)-> a.val-b.val);
        
        for (ListNode node : lists){
            if (node != null){
                minHeap.add(node);
            }
        }

        ListNode dummy = new ListNode(-1);
        ListNode current = dummy;

        while (!minHeap.isEmpty()){
            current.next = minHeap.poll();
            current = current.next;

            if (current.next != null){
                minHeap.add(current.next);
            }
        }

        return dummy.next;
    }
}